lib/fluent/plugin/out_file.rb

```diff
@@ -178,7 +178,7 @@ def format(tag, time, record)
     end
 
     def write(chunk)
-      path = extract_placeholders(@path_template, chunk.metadata)
+      path = extract_placeholders(@path_template.gsub(CHUNK_ID_PLACEHOLDER_PATTERN, dump_unique_id_hex(chunk.unique_id)), chunk.metadata)
```

Comment:
It seems better that the replacements of `CHUNK_ID_PLACEHOLDER_PATTERN` should be done by `extract_placeholders` instead of each plugins. For now, only these two plugins use this method, but it is useful if all plugins can use this feature automatically in the future.

Reply:
Yeah, it is one idea.
For it, we need to change `extract_placeholders` API.

1. 2nd argument changed to chunk, not `chunk.metadata`

For compatibility, we need to check 2nd argument type.

```
metadata = chunk.metadata if metadata.is_a?(Chunk)
```

2. Add 3rd argument for chunk

`def extract_placeholders(str, metadata, chunk = nil)`

`metadata` is from `chunk` so this API is a little strange.

3. Add 3rd argument for additional params

This avoids further API changes.

```
def extract_placeholders(str, metadata, extras = nil)
  if extras && extras.has_key?(:chunk)
  end
end
```

I think 1 is better because this doesn't break exsiting API and doesn't add more argument.
We can change Plugin API before stable announcement.

Reply:
👍 > 1

Reply:
I agree to change `extract_placeholders` API.
:+1: > 1 and 3. The both looks good to me.

Reply:
:+1:  > 1